src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+     */
+    protected void updateSamples()
+    {
+        // Split calculation of probe timers from sending probes for testability
+        calculateProbes(samples, dynamicLatencyProbeInterval);
+
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        schedulePings(samples);
+    }
+
+    /**
+     * This method mutates the passed AnnotatedMeasurements to implement capped exponential backoff per endpoint.
+     *
+     * The algorithm is as follows:
+     * 1. All samples get their millisSinceLastMeasure and millisSinceLastRequest fields
+     *    incremented by the passed interval
+     * 2. Any recently requested (ranked) endpoints that have not been measured recently (e.g. because the snitch
+     *    has sent them no traffic) get probes with exponential backoff.
+     *
+     * The backoff is capped at MAX_PROBE_INTERVAL_MS. Furthermore the probes are stopped after
+     * MAX_PROBE_INTERVAL_MS of no ranking requests as well.
+     *
+     * At the end of this method, any passed AnnotatedMeasurements that need latency probes will have non zero
+     * probeTimerMillis members set.
+     */
+    @VisibleForTesting
+    static void calculateProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples, long intervalMillis) {
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            if (entry.getKey().equals(FBUtilities.getBroadcastAddressAndPort()))
+                continue;
+
+            AnnotatedMeasurement measurement = entry.getValue();
+            long lastMeasure = measurement.millisSinceLastMeasure.getAndAdd(intervalMillis);
+            long lastRequest = measurement.millisSinceLastRequest.getAndAdd(intervalMillis);
+
+            if (lastMeasure >= MIN_PROBE_INTERVAL_MS && lastRequest < MAX_PROBE_INTERVAL_MS)
+            {
+                if (measurement.probeTimerMillis == 0)
+                {
+                    measurement.probeTimerMillis = intervalMillis;
+                }
+                else if (measurement.probeFuture != null && measurement.probeFuture.isDone())
+                {
+                    measurement.probeTimerMillis = Math.min(MAX_PROBE_INTERVAL_MS, measurement.probeTimerMillis * 2);
+                }
+            }
+            else
+            {
+                measurement.probeTimerMillis = 0;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    void schedulePings(Map<InetAddressAndPort, AnnotatedMeasurement> samples)
+    {
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            AnnotatedMeasurement measurement = entry.getValue();
+            long delay = measurement.probeTimerMillis;
+            long millisSinceLastRequest = measurement.millisSinceLastRequest.get();
+
+            if (millisSinceLastRequest > MAX_PROBE_INTERVAL_MS && !Gossiper.instance.isAlive(entry.getKey()))
```

Comment:
This way all of the global state (`gossiper`) and threading state (`latencyProbeExecutor`) are in this method instead of `calculateProbes`, which means that we can call `calculateProbes` without checking if the gossiper is active (to maintain the logical clock moving forward).

If we want to calculate the elapse time as you suggest instead of relying on the scheduler to schedule us ~1s then this becomes less of a problem.

Reply:
I think I like the simplicity of where things are right now and I want to take it back. It's still mostly bounded on how far off it can be from the configured value and it won't go too fast. That should be fine?